models/classes/resources/Service/InstanceMetadataCopier.php

```diff
@@ -24,15 +24,139 @@
 
+use helpers_File;
+use core_kernel_classes_Literal;
 use core_kernel_classes_Resource;
+use oat\generis\model\GenerisRdf;
+use core_kernel_classes_Property;
+use oat\generis\model\OntologyRdf;
+use oat\oatbox\filesystem\FileSystemService;
+use oat\generis\model\fileReference\FileReferenceSerializer;
+use oat\tao\model\resources\Contract\ClassMetadataMapperInterface;
 use oat\tao\model\resources\Contract\InstanceMetadataCopierInterface;
 
 class InstanceMetadataCopier implements InstanceMetadataCopierInterface
 {
+    /** @var ClassMetadataMapperInterface */
+    private $classMetadataMapper;
+
+    /** @var FileReferenceSerializer */
+    private $fileReferenceSerializer;
+
+    /** @var FileSystemService */
+    private $fileSystemService;
+
+    /** @var string[] */
+    private $blacklistedProperties = [
+        OntologyRdf::RDF_TYPE,
+    ];
+
+    /** @var array<string, core_kernel_classes_Property> */
+    private $sharedProperties = [];
+
+    /** @var array<string, bool> */
+    private $fileProperties = [];
+
+    public function __construct(
+        ClassMetadataMapperInterface $classMetadataMapper,
+        FileReferenceSerializer $fileReferenceSerializer,
+        FileSystemService $fileSystemService
+    ) {
+        $this->classMetadataMapper = $classMetadataMapper;
+        $this->fileReferenceSerializer = $fileReferenceSerializer;
+        $this->fileSystemService = $fileSystemService;
+    }
+
+    public function addPropertyUriToBlacklist(string $propertyUri): void
+    {
+        if (!in_array($propertyUri, $this->blacklistedProperties, true)) {
+            $this->blacklistedProperties[] = $propertyUri;
+        }
+    }
+
     public function copy(
         core_kernel_classes_Resource $instance,
         core_kernel_classes_Resource $destinationInstance
     ): void {
-        // TODO: Implement copy() method.
+        $originalClass = current($instance->getTypes());
+        $destinationClass = current($destinationInstance->getTypes());
+
+        $destinationProperties = $destinationClass->getProperties(true);
+        $this->sharedProperties = array_merge(
+            $this->sharedProperties,
+            array_intersect_key($originalClass->getProperties(), $destinationProperties)
+        );
+
+        foreach ($destinationProperties as $destinationProperty) {
+            $originalProperty = $this->getOriginalProperty($destinationProperty);
+
+            if (
+                $originalProperty === null
+                || in_array($originalProperty->getUri(), $this->blacklistedProperties, true)
+            ) {
+                continue;
+            }
+
+            $propertyValues = $instance->getPropertyValuesCollection($originalProperty);
+
+            /** @var core_kernel_classes_Literal|core_kernel_classes_Resource|core_kernel_classes_Resource[] $propertyValue */
+            foreach ($propertyValues->getIterator() as $propertyValue) {
+                if ($this->isFileProperty($originalProperty)) {
```

Comment:
Probably it will never reach this line, since we do not have File property working. How did you manage to test that?

Reply:
Idk
I just copied the logic we use to copy an instance.

Reply:
I see 😢. I would at least remove the unnecessary cache and move this method outside the loop and in the loop have only:

```suggestion
                if ($isFileProperty) {
```

As I mentioned here https://github.com/oat-sa/tao-core/pull/3526/files/e2dc74f8f7bbaca07e04fb264e75c268ff48f587#r874410844

Wdyt?